static_frame/core/loc_map.py

```diff
@@ -413,13 +421,107 @@ def loc_to_iloc(self: _HLMap,
+            ) -> np.ndarray:
+        '''
+        Given an encoding, unpack it into its constituent parts
+
+        Ex:
+            bit_offset_encoders = [0, 2, 4]
+
+            Encodings:
+                36  => [0, 4, 32] => [0, 1, 2]
+                 8  => [0, 8,  0] => [0, 2, 0]
+                10  => [2, 8,  0] => [2, 2, 0]
+                17  => [1, 0, 16] => [1, 0, 1]
+
+            Step 1:
+            Expand bit_offset_encoders into something more helpful -> masks
+
+            [0, 2, 4] is where the bit offsets start. They end one bit before the next offset.
+            Thus, the bit offset ends are:
+            [1, 3, 64] # since 64 is the max bit offset
+
+            From here, we build up a list of masks that each have the correct number of up bits
+
+            0 => [11] # 2 bit mask
+            2 => [11] # 2 bit mask
+            4 => [11] # 2 bit mask
+
+            Now, for each component (i.e. the number of bit_offset_encoders), we
+            apply to corresponding mask to the values AFTER they have been shifted backwards
+
+            36 == [10 01 00]
+             8 == [00 10 00]
+            10 == [00 10 10]
+            17 == [01 00 01]
+
+            Depth 0:
+                offset = bit_offset_encoders[0] = 0
+                36 => ([10 01 00] << 0) => [10 01 00] & [11] => [00 00 00] => 0
+                 8 => ([00 10 00] << 0) => [00 10 00] & [11] => [00 00 00] => 0
+                10 => ([00 10 10] << 0) => [00 10 10] & [11] => [00 00 10] => 2
+                17 => ([01 00 01] << 0) => [01 00 01] & [11] => [00 00 01] => 1
+
+            Depth 1:
+                offset = bit_offset_encoders[1] = 2
+                36 => ([10 01 00] << 2) => [10 01] & [11] => [00 01] => 1
+                 8 => ([00 10 00] << 2) => [00 10] & [11] => [00 10] => 2
+                10 => ([00 10 10] << 2) => [00 10] & [11] => [00 10] => 2
+                17 => ([01 00 01] << 2) => [01 00] & [11] => [00 00] => 0
+
+            Depth 2:
+                offset = bit_offset_encoders[2] = 4
+                36 => ([10 01 00] << 4) => [10] & [11] => [10] => 2
+                 8 => ([00 10 00] << 4) => [00] & [11] => [00] => 0
+                10 => ([00 10 10] << 4) => [00] & [11] => [00] => 0
+                17 => ([01 00 01] << 4) => [01] & [11] => [01] => 1
+
+            Result:
+                36 => [0, 1, 2]
+                 8 => [0, 2, 0]
+                10 => [2, 2, 0]
+                17 => [1, 0, 1]
+
+            NOTE: This is the inverse of the documentation in `build_encoded_indexers_map`
+        '''
+        assert bit_offset_encoders.dtype == np.uint64
+        assert bit_offset_encoders[0] == 0 # By definition, the first offset starts at 0!
+        assert encoded_arr.ndim == 1 # Encodings are always 1D
+
+        dtype = object if encoding_can_overflow else np.uint64
+
+        starts = bit_offset_encoders
+        stops = np.concatenate((starts[1:], np.array([64], dtype=dtype)))
```

Comment:
I wonder if there is a better way to do this: perhaps just:
```suggestion
        stops = np.roll(starts, -1)
        stops[-1] = 64
```

Reply:
While `roll` is slightly faster at large sizes, concatenate is about 5x faster for the typical use case here, which `len(starts)` will typically be between 2 - 5, since it only contains a single integer for each depth level.

If you don't mind, I'll leave it as is

Reply:
That is super counterintuitive! I guess the small size makes this a special case, but I still explored a few more options:

```python
>>> a1 = np.arange(5)   

>>> def a(): 
...     return np.concatenate((a1[1:], np.array([-1]))) 
...        

>>> def b(): 
...     x = np.roll(a1, -1) 
...     x[-1] = -1 
...     return x 
...     

>>> def c(): 
...     x = np.full(len(a1), -1) 
...     x[:-1] = a1[1:] 
...     return x 
...     

>>> def d(): 
...     x = np.empty(len(a1), dtype=a1.dtype) 
...     x[:-1] = a1[1:] 
...     x[-1] = -1 
...     return x 
...       

>>> %timeit a()                                                                                                                                 
2.2 µs ± 120 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit b()                                                                                                                                 
10.3 µs ± 647 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit c()                                                                                                                                 
2.35 µs ± 60.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit d()                                                                                                                                 
1.15 µs ± 21.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Looks like `d()` is the fastest at this scale, but I understand if you prefer your one-line example at the cost of ~1 additional µs.

Reply:
Nice! Let's do it haha